src/api/mod.rs

```diff
@@ -745,38 +790,83 @@ impl<T: Pixel> Context<T> {
+  /// then encoding can proceed. This is just a hint to the application, and
+  /// does not need to be called for encoding the second pass to work, so long
+  /// as the application continues to provide more data to [`twopass_in`] in a
+  /// loop until [`twopass_in`] returns `0`.
+  ///
+  /// [`twopass_in`]: #method.twopass_in
+  #[inline]
   pub fn twopass_bytes_needed(&mut self) -> usize {
     self.inner.rc_state.twopass_in(None).unwrap_or(0)
   }
 
-  /// Provide stats data produced in the first pass of a two-pass encode to the
-  /// second pass. On success this returns the number of bytes of that data
-  /// which were consumed. When encoding the second pass of a two-pass encode,
-  /// this should be called repeatedly in a loop before every call to
-  /// receive_packet() (including the very first one) until no bytes are
-  /// consumed, or until twopass_bytes_needed() returns 0.
+  /// Provides the stats data produced in the first pass of a two-pass encode
+  /// to the second pass.
+  ///
+  /// On success this returns the number of bytes of the data which were
+  /// consumed. When encoding the second pass of a two-pass encode, this should
+  /// be called repeatedly in a loop before every call to [`receive_packet`]
+  /// (including the very first one) until no bytes are consumed, or until
+  /// [`twopass_bytes_needed`] returns `0`.
+  ///
+  /// [`receive_packet`]: #method.receive_packet
+  /// [`twopass_bytes_needed`]: #method.twopass_bytes_needed
+  #[inline]
   pub fn twopass_in(&mut self, buf: &[u8]) -> Result<usize, EncoderStatus> {
     self.inner.rc_state.twopass_in(Some(buf)).or(Err(EncoderStatus::Failure))
   }
 
+  /// Returns the encoded data.
+  ///
+  /// # Examples
+  ///
+  /// ``` no_run
+  /// use rav1e::prelude::*;
+  /// #
+  /// # fn main() -> Result<(), EncoderStatus> {
+  /// let cfg = Config::default();
+  /// let mut ctx: Context<u8> = cfg.new_context();
+  /// let frame = ctx.new_frame();
+  ///
+  /// ctx.send_frame(frame)?;
+  /// ctx.flush();
+  ///
+  /// loop {
+  ///     match ctx.receive_packet() {
+  ///         Ok(packet) => { /* ... */ },
+  ///         Err(EncoderStatus::Encoded) => (),
+  ///         Err(EncoderStatus::LimitReached) => break,
+  ///         Err(err) => return Err(err),
+  ///     }
+  /// }
+  /// # Ok(())
+  /// # }
+  /// ```
+  #[inline]
   pub fn receive_packet(&mut self) -> Result<Packet<T>, EncoderStatus> {
     let inner = &mut self.inner;
     let pool = &mut self.pool;
 
     pool.install(|| inner.receive_packet())
   }
 
+  /// Flushes the encoder.
+  ///
+  /// Flushing signals the end of the video. After the encoder has been
+  /// flushed, no additional frames are accepted.
+  #[inline]
   pub fn flush(&mut self) {
     self.send_frame(None).unwrap();
   }
 
+  /// Produces a sequence header matching the current encoding context.
+  ///
+  /// Its format is compatible with the AV1 Matroska and ISOBMFF specification.
```

Comment:
You should add a note that it does not contain the needed config OBUs for e.g. color, which are required for some uses as part of the Matroska and ISOBMFF specifications, but are technically optional.

Reply:
Could you elaborate? I'm not familiar with this, I just copied the capi docs.

Reply:
As per: https://aomediacodec.github.io/av1-isobmff/#av1codecconfigurationbox

The last entry, `configOBUs` contains zero or more, well, config OBUs (sequence header with color config, for example). rav1e currently doesnt return any of these (so, zero) as part of this function. Other things, like libavcodec/libavformat put them in there.

They're needed for e.g. generating a portable MSE codec string for use in a DASH manifest.

Just think it's worth a note that we don't include them.

Reply:
> > `make_mut` works fine for the purpose.
> 
> Eh, I'm not a big fan of this. Us returning an `Arc` kind of implies that we might've stored a reference ourselves, in which case `make_mut` making a clone of the underlying value would potentially break us. It just feels like adding un-clarity to the API for no particular reason.
> 

We could add explicit setters/getters.

> Adding an `IntoFrame` impl for `Frame` would make much more sense in my opinion.
> 

We could add it easily, but you have to support Arc<Frames> from the users as well anyway.

> > Ideally we could rework the whole thing so that the Frame is pool-backed and refcounted without an external Arc
> 
> Any particular reason for this over the regular `Arc`?

Not many beside maybe a different ergonomy compared to what Arc provides.

I'm not against adding the `Frame` variant for IntoFrame and simplify the frame factory since it is not pool backed right now and probably it won't be in a long while.

Reply:
@tdaede I mentioned it above:

> They're needed for e.g. generating a portable MSE codec string for use in a DASH manifest.

That is, the *full* MSE codec string, which makes sure color is set properly.